File: 841/main.go
package main

import "fmt"

func main() {
	rooms := [][]int{
		[]int{1, 3},
		[]int{3, 0, 1},
		[]int{2},
		[]int{0},
	}
	fmt.Println(canVisitAllRooms(rooms))
}

func canVisitAllRooms(rooms [][]int) bool {
	visited := []bool{}
	for i := 0; i < len(rooms); i++ {
		visited = append(visited, false)
	}
	queue := []int{0}
	for len(queue) != 0 {
		currRoom := queue[0]
		queue = queue[1:]
		for _, key := range rooms[currRoom] {
			if !visited[key] {
				queue = append(queue, key)
			}
		}
		visited[currRoom] = true
	}
	for _, val := range visited {
		if !val {
			return false
		}
	}
	return true
}

File: 841/main.py
from typing import List
from collections import deque


class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        """
        approach
        go to zero, get keys, put onto queue
        while queue:
            pop
            visit room
        """
        queue = deque(rooms[0])
        visited = set([0])
        while queue:
            room_to_visit = queue.popleft()
            visited.add(room_to_visit)
            for next_room in rooms[room_to_visit]:
                if next_room not in visited:
                    queue.append(next_room)
        return len(visited) == len(rooms)
